### PR TITLE
DAOS-12567 control,engine: Fix list-devices for MD-on-SSD

### PR DESCRIPTION
**Pull request by tanabarr**
Update dmg storage query list-devices to indicate which roles each
device is currently performing. Show unique list of target IDs and if
a system XStream is running on the device.

Required-githooks: true

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by tanabarr** [comment]
Output of `dmg storage query list-devices` with PR change:
```
localhost
---------
  Devices
    UUID:e068f731-bbcb-4c71-9fc6-067e400115f7 [TrAddr:0000:2c:00.0]
      Roles:wal SysXS Targets:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15] Rank:0 State:NORMAL LED:NA
    UUID:93620668-40d0-40bd-80bd-709773e765fb [TrAddr:0000:2d:00.0]
      Roles:data,meta SysXS Targets:[0 6 12] Rank:0 State:NORMAL LED:NA
    UUID:6d00150b-6558-44be-a44a-651a39bc18f4 [TrAddr:0000:2e:00.0]
      Roles:data,meta Targets:[1 7 13] Rank:0 State:NORMAL LED:NA
    UUID:0611eadc-1107-496c-9c9c-4365e971fbd7 [TrAddr:0000:63:00.0]
      Roles:data,meta Targets:[2 8 14] Rank:0 State:NORMAL LED:NA
    UUID:0992fdfb-5d1e-4a9e-a305-1a88512a8eaa [TrAddr:0000:64:00.0]
      Roles:data,meta Targets:[3 9 15] Rank:0 State:NORMAL LED:NA
    UUID:58b263e8-2638-4046-8bf1-24c6f965e83b [TrAddr:0000:65:00.0]
      Roles:data,meta Targets:[4 10] Rank:0 State:NORMAL LED:NA
    UUID:5966fc8d-6cb9-4ea2-8998-a5f35d9a9576 [TrAddr:0000:66:00.0]
      Roles:data,meta Targets:[5 11] Rank:0 State:NORMAL LED:NA
    UUID:69577b75-a995-4b1b-ba07-ee09e368b75f [TrAddr:0000:a3:00.0]
      Roles:wal SysXS Targets:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15] Rank:1 State:NORMAL LED:NA
    UUID:ddfd35a9-44f7-4816-b1a7-ef088038496c [TrAddr:0000:a4:00.0]
      Roles:data,meta SysXS Targets:[0 6 12] Rank:1 State:NORMAL LED:NA
    UUID:fcedc0d2-5f73-4e39-b20b-22c3d57626e1 [TrAddr:0000:a5:00.0]
      Roles:data,meta Targets:[1 7 13] Rank:1 State:NORMAL LED:NA
    UUID:0fa5ca32-7bc0-407a-aa97-9c3f055f59af [TrAddr:0000:de:00.0]
      Roles:data,meta Targets:[2 8 14] Rank:1 State:NORMAL LED:NA
    UUID:736fe0f6-e6fa-4c20-a84c-2b0706d83711 [TrAddr:0000:df:00.0]
      Roles:data,meta Targets:[3 9 15] Rank:1 State:NORMAL LED:NA
    UUID:14aaa2a1-b062-4edc-832c-0ce90ab69275 [TrAddr:0000:e0:00.0]
      Roles:data,meta Targets:[4 10] Rank:1 State:NORMAL LED:NA
    UUID:1bf82df2-ab46-42f5-a482-47c2ac0271b9 [TrAddr:0000:e1:00.0]
      Roles:data,meta Targets:[5 11] Rank:1 State:NORMAL LED:NA
```

File: src/include/daos_srv/bio.h [hunk 1]
```diff
@@ -354,6 +354,7 @@ struct bio_dev_info {
 	int		       *bdi_tgts;
 	char		       *bdi_traddr;
 	uint32_t		bdi_dev_type;	/* reserved */
+	uint32_t		bdi_dev_roles;	/* reserved */
```
Comment (kjacque) [inline src/include/daos_srv/bio.h:357]: Is it really reserved? seems like it's used in srv_query.c

**Review by kjacque — APPROVED** [review]
One minor nit, no other issues noted.

File: src/include/daos_srv/bio.h (hunk 1 above)
Comment (wangshilong) [reply src/include/daos_srv/bio.h:357]: I think it intent to replace bdi_dev_type with bdi_dev_roles here..

**Comment by wangshilong** [comment]
I hit following errors when running format? 

 wolf-150 pbin: code = 5 description = "json: cannot unmarshal string into Go struct field BdevTierProperties.Properties.DeviceRoles of type storage.BdevDeviceRoles

File: src/include/daos_srv/bio.h (hunk 1 above)
Comment (tanabarr) [reply src/include/daos_srv/bio.h:357]: I will update on the next re-push, or as part of a subsequent PR.

**Comment by tanabarr** [comment]
> I hit following errors when running format?
> 
> wolf-150 pbin: code = 5 description = "json: cannot unmarshal string into Go struct field BdevTierProperties.Properties.DeviceRoles of type storage.BdevDeviceRoles

I can't recreate this, when formatting I don't see this issue:
```
[tanabarr@wolf-226 daos]$ install-rocky/bin/dmg storage format -i
Format Summary:
  Hosts     SCM Devices NVMe Devices
  -----     ----------- ------------
  localhost 2           14
```
can you please recheck that you have the most recent version of the patch and you are updating your daos_server_helper binary with:
```
 1016  chmod a-x install-rocky/bin/daos_server_helper
 1017  source utils/sl/setup_local.sh
 1018  sudo -E utils/setup_daos_server_helper.sh
```

**Pull request merged**